multiproof.go

```diff
@@ -274,3 +278,55 @@ func (mp MultiProof) Equal(other MultiProof) bool {
 	}
 	return mp.D.Equal(&other.D)
 }
+
+func groupPolynomialsByEvaluationPoint(fs [][]fr.Element, powersOfR []fr.Element, zs []uint8) [common.VectorLength][]fr.Element {
+	workersAggregations := make(chan [common.VectorLength][]fr.Element)
+
+	numWorkers := runtime.NumCPU()
+	batchSize := (len(fs) + numWorkers - 1) / numWorkers
+	for i := 0; i < numWorkers; i++ {
+		go func(start, end int) {
+			if end > len(fs) {
+				end = len(fs)
+			}
+			var groupedFs [common.VectorLength][]fr.Element
+			for i := start; i < end; i++ {
+				z := zs[i]
+				if len(groupedFs[z]) == 0 {
+					groupedFs[z] = make([]fr.Element, common.VectorLength)
+				}
+
+				for j := 0; j < common.VectorLength; j++ {
+					var scaledEvaluation fr.Element
+					scaledEvaluation.Mul(&powersOfR[i], &fs[i][j])
+					groupedFs[z][j].Add(&groupedFs[z][j], &scaledEvaluation)
+				}
+			}
+			workersAggregations <- groupedFs
+		}(i*batchSize, (i+1)*batchSize)
+	}
+
+	// Each worker has computed its own aggregation. Now we aggregate the results.
+	// This is bounded to reducing a `numWorkers` sized array of `common.VectorLength` sized arrays.
+	var groupedFs [common.VectorLength][]fr.Element
+	for i := 0; i < numWorkers; i++ {
+		workerAggregation := <-workersAggregations
+		for z := range workerAggregation {
+			if len(workerAggregation[z]) == 0 {
+				continue
+			}
+			// If this is the first time we see this evaluation point, we initialize it
+			// reusing the worker result.
+			if groupedFs[z] == nil {
+				groupedFs[z] = workerAggregation[z]
+				continue
+			}
+			// If not, we aggregate the worker result with the previous result for this evaluation.
+			for j := 0; j < common.VectorLength; j++ {
+				groupedFs[z][j].Add(&groupedFs[z][j], &workerAggregation[z][j])
+			}
+		}
+	}
```

Comment:
The "main" goroutine is taking these aggregated polynomials by evaluation point, and doing the final aggregation.

It's basically a kind of "map-reduce" style parallelization. It doesn't have any other trick or similar. Everything should _feel_ quite expected/normal, if something looks hairy or tricky, it's probably a bug.

Note: considering this last step is aggregating `runtime.NumCPU()` grouped polynomials, I experimented with also parallelizing this. But didn't make much of a difference and it complicated code more than needed. So this last step is single threaded.